### DISCUSSION TITLE
Proposition: Complement Lifecycle reducers with tokens

### DISCUSSION THREAD

**Discussion by RabugenTom** [pointfreeco/swift-composable-architecture#402] [Ideas]
Hello,

A lifecycle reducer which handles `onAppear` and `onDisappear` actions for SwiftUI views is presented in the case studies shipping with TCA. This higher order reducer is very convenient as it composes nicely with `IfLetStore`: the state is indeed allowed to being `nil` when reducing `onDisappear` actions which is a (currently crashing on purpose) limitation of `optional()` stores.

For this reason, with this lifecycle reducer, `onAppear` and `onDisappear` reduction depends only on the the environment. I find this configuration too limiting as it impedes fine grained control of cancellations in an app with many similar features visible at the same time. I propose an extension of the current approach where a token which is bound to the effective lifetime of a SwiftUI view is provided alongside the environment. Combined with an identifier, it allows to initiate or cancel long running effects only for a specific view without resorting to the state (albeit possible, see in the example below), and without affecting similar views with similar reducers still running alongside the appearing/disappearing view.

To this end, the token is provided in the form of a `LifecycleTokenizer` value. This value can be called as a function to generate a token-dependant identifier from another more general identifier (in the form which is currently used in the case studies). This value also allows to automatically type-erase token values as well as tokenized identifiers.

Furthermore, the use of this tokenizer is optional, and can still send `viewstore.send(.onAppear)` if no token is needed. 
Only the arguments of the `onAppear` and `onDisappear` functions in the reducer are changed from `(Environment)` to `(LifecycleTokenizer, Environment)`. If no token is specified, the tokenizer still works as expected and returns an `AnyHashable` identifier which depends only on the provided identifier.

Here is a proposed modification of the lifecycle reducer:
```swift
public struct LifecycleTokenizer: Hashable {
  let token: AnyHashable
  init<H: Hashable>(_ hashable: H?) {
    self.token = hashable as AnyHashable
  }

  public func tokenize<ID: Hashable>(_ id: ID) -> AnyHashable {
    TokenizedIdentifier(token: token, id: id)
  }

  public func callAsFunction<ID: Hashable>(_ id: ID) -> AnyHashable {
    tokenize(id)
  }
  
  // This hashable container does not need to be exposed and is erased
  // to AnyHashable on call sites. The `id` could be type-erased to `AnyHashable`
  // too when initializing, keeping only a generic initializer, but I think it's best
  // to keep generics alive as long as possible. `tokenize(_:ID)` could also
  // returns a typed `TokenizedIdentifier` but I feel it clutters the user with
  // irrelevant info.
  internal struct TokenizedIdentifier<ID>: Hashable where ID: Hashable {
    let token: AnyHashable
    let id: ID
    init(token: AnyHashable, id: ID) {
      self.token = token
      self.id = id
    }
  }
}

public extension Reducer {
  func lifecycle(
    onAppear: @escaping (LifecycleTokenizer, Environment) -> Effect<Action, Never>,
    onDisappear: @escaping (LifecycleTokenizer, Environment) -> Effect<Never, Never>
  ) -> Reducer<State?, LifecycleAction<Action>, Environment> {
    return .init { state, lifecycleAction, environment in
      switch lifecycleAction {
      case .onAppear(let token):
        return onAppear(LifecycleTokenizer(token), environment)
          .map(LifecycleAction.action)
      case .onDisappear(let token):
        return onDisappear(LifecycleTokenizer(token), environment)
          .fireAndForget()
      case .action(let action):
        guard state != nil else { return .none }
        return self.run(&state!, action, environment)
          .map(LifecycleAction.action)
      }
    }
  }
}

public enum LifecycleAction<Action> {
  case onAppear(token: AnyHashable? = nil)
  case onDisappear(token: AnyHashable? = nil)
  case action(Action)

  // Syntactic sugar to call `.onAppear` instead of `.onAppear()` when the token is nil.
  public static var onAppear: LifecycleAction<Action> { .onAppear(token: nil) }
  public static var onDisappear: LifecycleAction<Action> { .onDisappear(token: nil) }
}
```
And here is a example of application:
```swift
struct FeatureState: Equatable {}
enum FeatureAction {
  case noop
}

struct FeatureEffectIdentifier: Hashable {}

let featureReducer = Reducer<FeatureState, FeatureAction, Void> {
  state, action, environment in
  .none
}
.lifecycle { tokenize, environment in
  // Generate a tokenized identifier from `FeatureEffectIdentifier`:
  let id = tokenize(FeatureEffectIdentifier()) 

  return Just(FeatureAction.noop)
    .delay(for: .seconds(10), scheduler: RunLoop.main)
    .setFailureType(to: Never.self)
    .eraseToEffect()
    .cancellable(id: id)
} onDisappear: { tokenize, environment in
  let id = tokenize(FeatureEffectIdentifier())
  return .cancel(id: id)
}

struct FeatureView: View {
  let store: Store<FeatureState, LifecycleAction<FeatureAction>>

  // Token storage during the view's lifetime
  @State private var token = UUID() 

  var body: some View {
    WithViewStore(store) { viewStore in
      Text("Hello, world!")
        .onAppear {
          // Here the token is purely managed by the view,
          // but it can also depends on the state and be
          // set once prior sending `onAppear` to the
          // ViewStore like: 
          // `token = viewStore.state.identifier`
          
          viewStore.send(.onAppear(token: token))
        }
        .onDisappear {
          viewStore.send(.onDisappear(token: token))
        }
    }
  }
}
```
What do you think about this? 
Is there a better way to setup these things?
Should I prepare a pull-request?
Thanks!

**Comment by lukeredpath** [comment]
Just calling out some previous discussion along these lines you might not have seen
https://forums.swift.org/t/how-can-the-composable-architecture-really-become-composable/43045/16

And also this which seems like a nicer approach overall to handling view lifecycle events:
https://github.com/pointfreeco/swift-composable-architecture/commit/dc4b49ddbe99587a4c81e6f910ad490ccd9e672e#diff-a1c96e882538e90ca6c6ab801541c93934582dd86a282f57a006b03b96d0621dR410-R434
